ui/src/message_popup/renderers/response_renderer.jsx

```diff
@@ -0,0 +1,194 @@
+/* @flow weak */
+import React from 'react';
+import _ from 'lodash';
+import VelocityTransitionGroup from "velocity-react/velocity-transition-group";
+
+import TextField from 'material-ui/TextField';
+import RaisedButton from 'material-ui/RaisedButton';
+import Snackbar from 'material-ui/Snackbar';
+
+import SetIntervalMixin from '../../helpers/set_interval_mixin.js';
+import FeedbackCard from '../feedback_card.jsx';
+const ONE_SECOND = 1000;
+
+// This response renderer supports responding with text, and revising the text.
+export default React.createClass({
+  displayName: 'ResponseRenderer',
+
+  mixins: [SetIntervalMixin],
+
+  propTypes: {
+    question: React.PropTypes.object.isRequired,
+    scaffolding: React.PropTypes.object.isRequired,
+    limitMs: React.PropTypes.number.isRequired,
+    onLogMessage: React.PropTypes.func.isRequired,
+    onResponseSubmitted: React.PropTypes.func.isRequired
+  },
+
+  getInitialState() {
+    return {
+      isRevising: false,
+      isDoneRevising: false,
+      didRevise: false,
+      initialResponseText: '',
+      finalResponseText: '',
+      elapsedMs: 0
+    };
+  },
+
+  componentDidMount() {
+    this.setInterval(this.onTimerTick, ONE_SECOND);
+  },
+  
+  logMessage(type) {
+    const {initialResponseText, finalResponseText, elapsedMs} = this.state;    
+    this.props.onLogMessage(type, {
+      initialResponseText,
+      finalResponseText,
+      elapsedMs
+    });
+  },
+
+  submitResponse() {
+    const {
+      initialResponseText,
+      finalResponseText,
+      elapsedMs,
+      didRevise
+    } = this.state;
+    this.props.onResponseSubmitted({
+      initialResponseText,
+      finalResponseText,
+      elapsedMs,
+      didRevise
+    });
+  },
+
+  onTimerTick() {
+    if (!this.state.isDoneRevising) {
+      this.setState({ elapsedMs: this.state.elapsedMs + ONE_SECOND });
+    }
+  },
+
+  onTextChanged(e) {
+    const initialResponseText = e.target.value;
+    this.setState({
+      initialResponseText,
+      finalResponseText: initialResponseText
+    });
+  },
+
+  // This transitions to either: (revising mode) or (done).
+  onRespondClicked() {
+    const {isDoneRevising} = this.state;    
+    const {scaffolding} = this.props;
+    const {examples} = this.props.question;
+
+    // Capture response either way
+    this.logMessage('message_popup_response');
+
+    // Transition to revising or to being done
+    if (scaffolding.helpType === 'feedback' && examples.length > 0 && !isDoneRevising) {
+      this.setState({ isRevising: true });
+    } else {
+      this.submitResponse();
+    }
```

Comment:
Found a bug in here in testing - previously it wasn't logging the response in solution mode.